spring-kafka/src/main/java/org/springframework/kafka/support/serializer/DelegatingByTopicSerializer.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright 2019-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.support.serializer;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Pattern;
+
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.serialization.Serializer;
+
+import org.springframework.core.log.LogAccessor;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.StringUtils;
+
+/**
+ * A {@link Serializer} that delegates to other serializers based on a topic pattern.
+ *
+ * @author Gary Russell
+ * @since 2.8
+ *
+ */
+public class DelegatingByTopicSerializer implements Serializer<Object> {
+
+	private static final LogAccessor LOGGER = new LogAccessor(DelegatingDeserializer.class);
+
+	/**
+	 * Name of the configuration property containing the serialization selector map for
+	 * values with format {@code selector:class,...}.
+	 */
+	public static final String VALUE_SERIALIZATION_TOPIC_CONFIG = "spring.kafka.value.serialization.bytopic.config";
+
+	/**
+	 * Name of the configuration property containing the serialization topic pattern map for
+	 * keys with format {@code pattern:class,...}.
+	 */
+	public static final String KEY_SERIALIZATION_TOPIC_CONFIG = "spring.kafka.key.serialization.bytopic.config";
+
+	/**
+	 * Set to false to make topic pattern matching case-insensitive.
+	 */
+	public static final String CASE_SENSITIVE = "spring.kafka.value.serialization.bytopic.case.insensitive";
```

Comment:
I wonder if we should have copies (references) for these constants from the deserializer.
Kinda thinking how end-users are going to search for similar constants when they configure deserializer...
(Might be an overhead on my side - don't mind 😄 )